### PR TITLE
api: add get_bucket_policy and set_bucket_policy.

### PR DESCRIPTION
**Pull request by balamurugana**
Fixes ##377

### CODE REVIEW

**Comment by harshavardhana** [comment]
what else is needed here @balamurugana ?

**Comment by harshavardhana** [comment]
> what else is needed here @balamurugana ?

Checking again what else is pending here to get this merged @balamurugana ?

**Comment by balamurugana** [comment]
:-1: There is an issue which delete policy.  Fixing now.

**Comment by balamurugana** [comment]
When running below sequence, there is a left over `s3:ListBucket` statement.
```bash
$ mc.set_bucket_policy(Policy.READ_ONLY, BUCKET, 'hello')
$ mc.set_bucket_policy(Policy.READ_WRITE, BUCKET, 'world')
$ mc.set_bucket_policy(Policy.READ_WRITE, BUCKET)
$ mc.set_bucket_policy(Policy.NONE, BUCKET)
```

Fixing it now.

**Comment by balamurugana** [comment]
Below policy JSON is set.

```python
mc.set_bucket_policy(Policy.READ_ONLY, "mybucket", "hello")
```

```json
{"Statement": [{"Action": "s3:GetBucketLocation",
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket",
                 "Sid": ""},
                {"Action": "s3:ListBucket",
                 "Condition": {"StringEquals": {"s3:prefix": "hello"}},
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket",
                 "Sid": ""},
                {"Action": "s3:GetObject",
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket/hello*",
                 "Sid": ""}],
 "Version": "2012-10-17"}
```

```python
mc.set_bucket_policy(Policy.READ_WRITE, "mybucket", "world")
```

```json
{"Statement": [{"Action": "s3:GetBucketLocation",
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket",
                 "Sid": ""},
                {"Action": "s3:GetObject",
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket/hello*",
                 "Sid": ""},
                {"Action": "s3:ListBucket",
                 "Condition": {"StringEquals": {"s3:prefix": ["world", "hello"]}},
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket",
                 "Sid": ""},
                {"Action": "s3:ListBucketMultipartUploads",
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket",
                 "Sid": ""},
                {"Action": ["s3:ListMultipartUploadParts",
                             "s3:GetObject",
                             "s3:AbortMultipartUpload",
                             "s3:DeleteObject",
                             "s3:PutObject"],
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket/world*",
                 "Sid": ""}],
 "Version": "2012-10-17"}
```

```python
mc.set_bucket_policy(Policy.READ_WRITE, "mybucket")
```

```json
{"Statement": [{"Action": "s3:GetBucketLocation",
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket",
                 "Sid": ""},
                {"Action": "s3:GetObject",
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket/hello*",
                 "Sid": ""},
                {"Action": "s3:ListBucketMultipartUploads",
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket",
                 "Sid": ""},
                {"Action": ["s3:ListMultipartUploadParts",
                             "s3:GetObject",
                             "s3:AbortMultipartUpload",
                             "s3:DeleteObject",
                             "s3:PutObject"],
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket/world*",
                 "Sid": ""},
                {"Action": "s3:ListBucket",
                 "Condition": {"StringEquals": {"s3:prefix": ["world", "hello"]}},
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket",
                 "Sid": ""},
                {"Action": "s3:ListBucket",
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket",
                 "Sid": ""},
                {"Action": ["s3:ListMultipartUploadParts",
                             "s3:GetObject",
                             "s3:AbortMultipartUpload",
                             "s3:DeleteObject",
                             "s3:PutObject"],
                 "Effect": "Allow",
                 "Principal": {"AWS": "*"},
                 "Resource": "arn:aws:s3:::mybucket/*",
                 "Sid": ""}],
 "Version": "2012-10-17"}
```

File: minio/api.py [hunk 1]
```diff
@@ -302,6 +304,81 @@ def remove_bucket(self, bucket_name):
...
+        Set bucket policy of given bucket name and object prefix.
+
+        :param bucket_name: Bucket name.
+        :param object_prefix: Object prefix.
+        """
+        is_valid_bucket_name(bucket_name)
+
+        policy_dict = self._get_bucket_policy(bucket_name)
+        if policy_access == policy.Policy.NONE and not policy_dict:
+            return
```
Comment (harshavardhana) [inline minio/api.py]: We need to return an error here..We raise an exception. `NoSuchBucketPolicy`

File: minio/policy.py [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+minio.policy
+
+This module implements policy management.
+
+:copyright: (c) 2015 by Minio, Inc.
```
Comment (harshavardhana) [inline minio/policy.py]: copyright should be 2016

File: minio/policy.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
+# -*- coding: utf-8 -*-
+# Minio Python Library for Amazon S3 Compatible Cloud Storage
+# (C) 2015 Minio, Inc.
```
Comment (harshavardhana) [inline minio/policy.py]: Should be 2016.

File: minio/policy.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+
+
+# Policy Enums.
+class Policy:
+    NONE = "none"
+    READ_ONLY = "readonly"
+    READ_WRITE = "readwrite"
+    WRITE_ONLY = "writeonly"
+
+
```
Comment (harshavardhana) [inline minio/policy.py]: All these private functions require comments @balamurugana

File: minio/policy.py [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+    return policy
+
+
+def set_policy(statements, policy, bucket_name, prefix=''):
+    out = _remove_statements(statements, Policy.READ_WRITE,
+                             bucket_name, prefix)
+    ns = _new_statements(policy, bucket_name, prefix)
+    append_statements(out, ns)
+
+    return out
```
Comment (harshavardhana) [inline minio/policy.py]: Everything else looks good apart from the comments.

File: minio/policy.py (hunk 2 above)
Comment (balamurugana) [inline minio/policy.py]: Done

File: minio/policy.py (hunk 3 above)
Comment (balamurugana) [inline minio/policy.py]: Done

File: minio/policy.py (hunk 4 above)
Comment (balamurugana) [inline minio/policy.py]: Done

**Comment by harshavardhana** [comment]
@balamurugana can you add  examples, update README.md and API.md as well.

**Comment by balamurugana** [comment]
> @balamurugana can you add examples, update README.md and API.md as well.

Done in latest refresh.

**Pull request merged**